File: C/Questao1_ProdutorConsumidor.c
#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>
#include <semaphore.h>
#include <unistd.h>

#define BUFFER_SIZE 5
#define NUM_ITENS 10

int buffer[BUFFER_SIZE];
int in = 0;
int out = 0;

// Semáforos
sem_t cheio;
sem_t vazio;
pthread_mutex_t mutex;

void* produtor(void* arg) {
    for (int i = 0; i < NUM_ITENS; i++) {
        int item = rand() % 100;

        sem_wait(&vazio);
        pthread_mutex_lock(&mutex);

        buffer[in] = item;
        printf("Produtor produziu: %d na posição %d\n", item, in);
        in = (in + 1) % BUFFER_SIZE;

        pthread_mutex_unlock(&mutex);
        sem_post(&cheio);

        sleep(1);  // Simula tempo de produção
    }
    pthread_exit(NULL);
}

void* consumidor(void* arg) {
    for (int i = 0; i < NUM_ITENS; i++) {
        sem_wait(&cheio);
        pthread_mutex_lock(&mutex);

        int item = buffer[out];
        printf("Consumidor consumiu: %d da posição %d\n", item, out);
        out = (out + 1) % BUFFER_SIZE;

        pthread_mutex_unlock(&mutex);
        sem_post(&vazio);

        sleep(2);  // Simula tempo de consumo
    }
    pthread_exit(NULL);
}

int main() {
    pthread_t tprodutor, tconsumidor;

    sem_init(&cheio, 0, 0);
    sem_init(&vazio, 0, BUFFER_SIZE);
    pthread_mutex_init(&mutex, NULL);

    pthread_create(&tprodutor, NULL, produtor, NULL);
    pthread_create(&tconsumidor, NULL, consumidor, NULL);

    pthread_join(tprodutor, NULL);
    pthread_join(tconsumidor, NULL);

    sem_destroy(&cheio);
    sem_destroy(&vazio);
    pthread_mutex_destroy(&mutex);

    return 0;
}

File: Java/Questao1_ProdutorConsumidor.java
import java.util.concurrent.Semaphore;

public class Questao1_ProdutorConsumidor {
    static final int BUFFER_SIZE = 5;
    static int[] buffer = new int[BUFFER_SIZE];
    static int in = 0, out = 0;

    static Semaphore cheio = new Semaphore(0);
    static Semaphore vazio = new Semaphore(BUFFER_SIZE);
    static Semaphore mutex = new Semaphore(1);

    static class Produtor extends Thread {
        public void run() {
            try {
                for (int i = 0; i < 10; i++) {
                    int item = (int) (Math.random() * 100);
                    vazio.acquire();
                    mutex.acquire();

                    buffer[in] = item;
                    System.out.println("Produtor produziu: " + item + " na posição " + in);
                    in = (in + 1) % BUFFER_SIZE;

                    mutex.release();
                    cheio.release();

                    Thread.sleep(1000);
                }
            } catch (InterruptedException e) {}
        }
    }

    static class Consumidor extends Thread {
        public void run() {
            try {
                for (int i = 0; i < 10; i++) {
                    cheio.acquire();
                    mutex.acquire();

                    int item = buffer[out];
                    System.out.println("Consumidor consumiu: " + item + " da posição " + out);
                    out = (out + 1) % BUFFER_SIZE;

                    mutex.release();
                    vazio.release();

                    Thread.sleep(1500);
                }
            } catch (InterruptedException e) {}
        }
    }

    public static void main(String[] args) {
        new Produtor().start();
        new Consumidor().start();
    }
}
